### PR TITLE
Add support regexps for String#start_with?

### PR DESCRIPTION
**Pull request by zhublik**

### CODE REVIEW

**Comment by zhublik** [comment]
I have signed the OCA.

File: src/main/ruby/truffleruby/core/string.rb [hunk 1]
```diff
@@ -1490,12 +1490,22 @@ def rindex(sub, finish=undefined)
   end
 
   def start_with?(*prefixes)
+    binding = TrufflePrimitive.caller_binding
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/string.rb]: Getting the caller binding can be expensive, could you get it only for the case where it's needed?
In this case, an easy way seems to be:
```ruby
  def start_with?(*prefixes)
    binding = TrufflePrimitive.caller_binding if prefixes.any?(Regexp)
```

File: src/main/ruby/truffleruby/core/string.rb [hunk 2]
```diff
@@ -1490,12 +1490,22 @@ def rindex(sub, finish=undefined)
...
+
     prefixes.each do |original_prefix|
-      prefix = Truffle::Type.rb_check_convert_type original_prefix, String, :to_str
-      unless prefix
-        raise TypeError, "no implicit conversion of #{original_prefix.class} into String"
+      case original_prefix
+      when Regexp
+        Truffle::Type.compatible_encoding(self, original_prefix)
+        match_data = original_prefix.match_onwards(self, 0, true)
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/string.rb]: @aardvark179 Can you double check this sounds right for `String#start_with?(Regexp)`?
It sounds good to me, but we have quite a few not-so-clearly-named Regexp matching primitives.

**Comment by eregon** [comment]
@zhublik Thank you for the PR!
The graalvmbot should react soon regarding the OCA, I can see your name in the list.

File: src/main/ruby/truffleruby/core/string.rb (hunk 1 above)
Comment (eregon) [reply src/main/ruby/truffleruby/core/string.rb]: I'd simplify and just get the binding inside the block, but unfortunately `TrufflePrimitive.caller_binding` does not work inside blocks yet.

File: src/main/ruby/truffleruby/core/string.rb (hunk 2 above)
Comment (aardvark179) [reply src/main/ruby/truffleruby/core/string.rb]: Yes, that looks like the correct primitive with the correct arguments for doing this.

File: src/main/ruby/truffleruby/core/string.rb (hunk 1 above)
Comment (zhublik) [reply src/main/ruby/truffleruby/core/string.rb]: @eregon, i've added the changes here: b4316a0

**Comment by eregon** [comment]
Merged in 3ff049520755d4a6d9a1edbae92d2c2991bb6228

**Pull request merged**